cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1128 @@
+  ,
+  ublkcp
+#endif // _CUB_HAS_TRANSFORM_UBLKCP
+};
+
+// TODO(bgruber): only needed so we can instantiate the kernel generically for any policy. Remove when fallback_for is
+// dropped.
+template <typename, typename Offset, typename F, typename RandomAccessIteartorOut, typename... RandomAccessIteartorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::fallback_for>,
+  Offset,
+  int,
+  F,
+  RandomAccessIteartorOut,
+  RandomAccessIteartorIn...)
+{}
+
+template <int BlockThreads>
+struct prefetch_policy_t
+{
+  static constexpr int BLOCK_THREADS = BlockThreads;
+  // items per tile are determined at runtime. these (inclusive) bounds allow overriding that value via a tuning policy
+  static constexpr int ITEMS_PER_THREAD_NO_INPUT = 2; // items per thread when no input streams exits (just filling)
+  static constexpr int MIN_ITEMS_PER_THREAD      = 1;
+  static constexpr int MAX_ITEMS_PER_THREAD      = 32;
+};
+
+// Prefetches (at least on Hopper) a 128 byte cache line. Prefetching out-of-bounds addresses has no side effects
+template <typename T>
+_CCCL_DEVICE void prefetch(const T* addr)
+{
+  // TODO(bgruber): prefetch to L1 may be even better
+  asm volatile("prefetch.L2 [%0];" : : "l"(addr) : "memory");
+}
+
+// TODO(bgruber): there is also the cp.async.bulk.prefetch instruction available on Hopper. May improve perf a tiny bit
+// as we need to create less instructions to prefetch the same amount of data.
+
+// overload for any iterator that is not a pointer, do nothing
+template <typename It, ::cuda::std::__enable_if_t<!::cuda::std::is_pointer<It>::value, int> = 0>
+_CCCL_DEVICE void prefetch(It)
+{}
+
+// this kernel guarantees stable addresses for the parameters of the user provided function
+template <typename, typename Offset, typename F, typename RandomAccessIteartorOut, typename... RandomAccessIteartorIn>
+_CCCL_DEVICE void transform_kernel_impl(
+  ::cuda::std::integral_constant<Algorithm, Algorithm::prefetch>,
+  Offset num_items,
+  int num_elem_per_thread,
+  F f,
+  RandomAccessIteartorOut out,
+  RandomAccessIteartorIn... ins)
+{
+  const int tile_size = blockDim.x * num_elem_per_thread;
+  const Offset offset = static_cast<Offset>(blockIdx.x) * tile_size;
+
+  for (int j = 0; j < num_elem_per_thread; ++j)
+  {
+    // TODO(bgruber): replace by fold over comma in C++17
+    const auto idx = offset + (j * blockDim.x + threadIdx.x);
+    int dummy[]    = {(prefetch(ins + idx), 0)..., 0}; // extra zero to handle empty packs
+    (void) &dummy; // nvcc 11.1 needs extra strong unused warning supression
+    (void) &idx; // nvcc 11.1 needs extra strong unused warning supression
+  }
+
+  // ahendriksen: various unrolling yields less <1% gains at much higher compile-time cost
+  // TODO(bgruber): A6000 disagrees
+#pragma unroll 1
+  for (int j = 0; j < num_elem_per_thread; ++j)
+  {
+    const auto idx = offset + (j * blockDim.x + threadIdx.x);
+    if (idx < num_items)
```

Comment:
if `offset >= 0`, `num_items >= 0`, `offset < num_items` and then we could write `num_items - offset` outside the loop

Reply:
for `threadIdx.x` is tricky due to the conversion

Reply:
I don't understand what you are suggesting. The conditions you stated are true, but there is no `num_items - offset` I can hoist out of the loop? Please elaborate! Thx!

Reply:
Given a later comment, I understand what you mean now:
```c++
    num_items -= offset;
    (ins += offset,...);
    out += offset;
```

Reply:
it is ok to move `threadIdx.x` as you suggested. My only concern is when `offset/num_items` are unsigned integer (`threadIdx.x` can be converted to a signed integer) and `offset + threadIdx.x > num_items`. (see my previous comment related to the line `num_items -= offset;`)

Reply:
@elstehle obviated because I use `idx < tile_size` in all kernels now.